lightning/src/ln/interactivetxs.rs

```diff
@@ -349,55 +557,32 @@ impl NegotiationContext {
 			return Err(AbortReason::ExceededNumberOfInputsOrOutputs);
 		}
 
-		// TODO: How do we enforce their fees cover the witness without knowing its expected length?
-		const INPUT_WEIGHT: u64 = BASE_INPUT_WEIGHT + EMPTY_SCRIPT_SIG_WEIGHT;
-
 		// - the peer's paid feerate does not meet or exceed the agreed feerate (based on the minimum fee).
-		let counterparty_output_weight_contributed: u64 = self
-			.counterparty_outputs_contributed()
-			.map(|output| {
-				(8 /* value */ + output.script_pubkey.consensus_encode(&mut sink()).unwrap() as u64)
-					* WITNESS_SCALE_FACTOR as u64
-			})
-			.sum();
-		let counterparty_weight_contributed = counterparty_output_weight_contributed
-			+ self.counterparty_inputs_contributed().count() as u64 * INPUT_WEIGHT;
-		let counterparty_fees_contributed =
-			counterparty_inputs_value.saturating_sub(counterparty_outputs_value);
-		let mut required_counterparty_contribution_fee =
-			fee_for_weight(self.feerate_sat_per_kw, counterparty_weight_contributed);
-		if !self.holder_is_initiator {
-			// if is the non-initiator:
-			// 	- the initiator's fees do not cover the common fields (version, segwit marker + flag,
-			// 		input count, output count, locktime)
-			let tx_common_fields_weight =
-		        (4 /* version */ + 4 /* locktime */ + 1 /* input count */ + 1 /* output count */) *
-		            WITNESS_SCALE_FACTOR as u64 + 2 /* segwit marker + flag */;
-			let tx_common_fields_fee =
-				fee_for_weight(self.feerate_sat_per_kw, tx_common_fields_weight);
-			required_counterparty_contribution_fee += tx_common_fields_fee;
-		}
-		if counterparty_fees_contributed < required_counterparty_contribution_fee {
-			return Err(AbortReason::InsufficientFees);
-		}
+		self.check_counterparty_fees(remote_inputs_value.saturating_sub(remote_outputs_value))?;
 
-		// Inputs and outputs must be sorted by serial_id
-		let mut inputs = self.inputs.into_iter().collect::<Vec<_>>();
-		let mut outputs = self.outputs.into_iter().collect::<Vec<_>>();
-		inputs.sort_unstable_by_key(|(serial_id, _)| *serial_id);
-		outputs.sort_unstable_by_key(|(serial_id, _)| *serial_id);
+		let holder_is_initiator = self.holder_is_initiator;
+		let (remote_inputs, local_inputs): (Vec<_>, Vec<_>) =
+			self.inputs.into_values().partition(|InteractiveTxInput { serial_id, .. }| {
+				is_serial_id_valid_for_counterparty(holder_is_initiator, serial_id)
+			});
+		let (remote_outputs, local_outputs): (Vec<_>, Vec<_>) =
+			self.outputs.into_values().partition(|InteractiveTxOutput { serial_id, .. }| {
+				is_serial_id_valid_for_counterparty(holder_is_initiator, serial_id)
+			});
```

Comment:
Why partition here just to recombine later? Could we use `filter` inside `ConstructedTransaction` instead to avoid four additional `Vec` allocations?